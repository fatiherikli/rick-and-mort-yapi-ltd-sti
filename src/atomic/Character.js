import React, { Component } from 'react';
import Error from '../components/Error';
import Image from '../components/Image';
import Tag from '../components/Tag';

import { setState, getState } from './store';

const TAGS = [
    ['name', 'name'],
    ['status', 'status'],
    ['species', 'species'],
    ['gender', 'gender'],
    ['origin', 'origin.name'],
];

export default class Character extends Component {
    state = {
        loading: true,
        error: null,
        response: null,
    }

    componentDidMount() {
        if (this.props.autoload) {
            const { uri } = this.props;
            if (getState(uri)) {
                this.setState({
                    response: getState(uri),
                    error: null,
                    loading: false,
                });
            } else {
                fetch(
                    this.props.uri
                ).then(
                    response => response.json()
                ).then(
                    this.handleFetch
                ).catch(
                    this.handleError
                )
            }
        }
    }

    handleFetch = response => {
        this.setState({
            loading: false,
            error: null,
            response,
        });

        setState({
            [this.props.uri]: response,
        });
    }

    handleError = error => {
        this.setState({
            loading: false,
            response: null,
            error: String(error),
        });
    }

    render() {
        const { loading, error, response } = this.state;
        return (
          <li>
            { error && <Error error={ error } />}
            { loading ? (
                <Image
                    src={ 'https://loading.io/spinners/interwind/lg.ball-interwind-preloader.gif' }
                    size={ 150 }
                />
            ) : (
                <div style={{ overflow: 'hidden' }}>
                    <div style={{ float: 'left' }}>
                        <Image
                            src={ response.image }
                            size={ 150 }
                        />
                    </div>
                    <div style={{ float: 'left', width: '7em', marginBottom: '1em' }}>
                        {
                            TAGS.map(
                                ([label, path]) => 
                                    <Tag
                                        character={ response }
                                        label={ label }
                                        key={ path }
                                        path={ path }
                                    />
                            )
                        }
                    </div>
                </div>
            ) }
          </li>
        );
    }
}

Character.defaultProps = {
    autoload: true,
};
