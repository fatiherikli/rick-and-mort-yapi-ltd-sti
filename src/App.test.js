import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Home from './pages/home';
import Episode from './pages/episode';
import Character from './atomic/Character';

configure({ adapter: new Adapter() });

describe('Pages', () => {
  it('Home', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ { pathname: '/', key: 'testKey' } ]}>
        <Home autoload={ false } />
      </MemoryRouter>
    );
    
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Detail', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ { pathname: '/', key: 'testKey' } ]}>
        <Episode autoload={ false } />
      </MemoryRouter>
    );
    
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('Atomic', () => {
  it('Character', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ { pathname: '/', key: 'testKey' } ]}>
        <Character autoload={ false } />
      </MemoryRouter>
    );
    
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});