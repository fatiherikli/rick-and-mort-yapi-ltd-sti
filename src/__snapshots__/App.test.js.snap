// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atomic Character 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Character
      autoload={false}
    >
      <li>
        <Image
          size={150}
          src="https://loading.io/spinners/interwind/lg.ball-interwind-preloader.gif"
        >
          <div
            style={
              Object {
                "background": "url(https://loading.io/spinners/interwind/lg.ball-interwind-preloader.gif)",
                "backgroundSize": 75,
                "height": 150,
                "width": 150,
              }
            }
          >
            <img
              alt="SOLENYA!"
              src="https://loading.io/spinners/interwind/lg.ball-interwind-preloader.gif"
              style={
                Object {
                  "height": 150,
                  "width": 150,
                }
              }
              width={150}
            />
          </div>
        </Image>
      </li>
    </Character>
  </Router>
</MemoryRouter>
`;

exports[`Pages Detail 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Episode
      autoload={false}
    >
      <div>
        <Loading>
          <ul>
            <li>
              <Link
                className="back"
                replace={false}
                to="/"
              >
                <a
                  className="back"
                  href="/"
                  onClick={[Function]}
                >
                  ⎈
                </a>
              </Link>
            </li>
          </ul>
        </Loading>
      </div>
    </Episode>
  </Router>
</MemoryRouter>
`;

exports[`Pages Home 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Home
      autoload={false}
    >
      <div>
        <Loading>
          <ul>
            <li>
              <Link
                className="back"
                replace={false}
                to="/"
              >
                <a
                  className="back"
                  href="/"
                  onClick={[Function]}
                >
                  ⎈
                </a>
              </Link>
            </li>
          </ul>
        </Loading>
      </div>
    </Home>
  </Router>
</MemoryRouter>
`;
